MaxEnt.py

```diff
@@ -1,181 +1,122 @@
+        sorted_pairs = sorted(
+            name_weight_pair.items(), key=lambda x: x[1], reverse=True
+        )
 
-            for c in range(len(cand[t])):
-                writer.writerow([cand[t][c], f"{round(P[c]*100, 1)}%"])
+        return sorted_pairs
+
+    def compute_probabilities(self, violations):
+        """Predicts probabilities for a particular violation profile.
+
+        Args:
+            violations (np.array[np.array[int]]): the violation profiles of a list of tableaux.
+
+        Returns:
+            np.array[np.array[float]]: predicts probability for each candidate of given URs.
+        """
+        harmonies = np.dot(violations, self.w)
+        exp = np.exp(-harmonies)
+        marginalization = np.sum(exp, axis=0)
+        return exp / marginalization
+
+    def SGD_learn(self, candidates, violations, srs, *regs, iters=10000, eta=0.05):
+        """Updates constraint weights using SGD.
+
+        Args:
+            candidates (np.array[np.array[str]]): candidates considered for each UR.
+            violations (np.array[np.array[int]]): the violation profiles of a list of tableaux.
+            srs (np.array[np.array[float]]): probability of a surface form observed for a given UR.
+            *regs: a variable number of regularization terms. Can be None.
+            iters (int): how many iterations to perform. Defaults to 10000.
+            eta (float): learning rate. Defaults to 0.05.
+
+        Returns:
+            _type_: _description_
+        """
+        # Initialization: begins with the weights that are currently stored.
+        w = self.w
+
+        # Keeps a history of the change in weights.
+        history = [w.squeeze()]
+
+        # Passes through all the training data in each iteration,
+        for _ in range(iters):
+            # Generates a random array of indices of the shape (1, length |URs|).
+            training_order = np.random.choice(
+                a=range(len(violations)), size=len(violations), replace=False
+            )
 
+            dL_dw = 0
 
-# /* --------------------------------------------------------------------------- */
-# /*                 Applies to a toy dataset                                    */
-# /* --------------------------------------------------------------------------- */
-if __name__ == "__main__":
-    from OTSoft_file_reader import get_info
+            # Loops through each tableau in the dataset by random order.
+            for t in training_order:
+                # Skips the tableaux without a winning candidate.
+                if srs[t].count(None) == len(srs[t]):
+                    continue
 
-    constraints, urs, candidates, violations, wIdx = get_info(
-        "HayesPseudoKorean-RichBase.txt"
-    )
+                # Violation profile of the winning candidate(s).
+                winner_violations = np.zeros(len(self.const))
+                for i, val in enumerate(srs[t]):
+                    if val:
```

Comment:
We should be able to get the weighted violations using vectorized operations instead of looping, something like:

```
probabilities = srs[t]
tableau = violations[t]
winner_violations = (probabilities * tableau).sum(axis=0)
```

I didn't actually test this so the syntax / dimensionality might be off.